staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -333,54 +334,86 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 		return r.base.RoundTrip(req)
 	}
 
-	creds, err := r.a.getCreds()
+	creds, _, err := r.a.getCreds()
 	if err != nil {
 		return nil, fmt.Errorf("getting credentials: %v", err)
 	}
 	if creds.token != "" {
+		req = utilnet.CloneRequest(req)
 		req.Header.Set("Authorization", "Bearer "+creds.token)
 	}
 
+	// make it possible for r.a.cert to detect when creds have already been fetched
+	ctx := context.WithValue(req.Context(), credsKey, creds)
+	req = req.WithContext(ctx)
+
 	res, err := r.base.RoundTrip(req)
 	if err != nil {
 		return nil, err
 	}
 	if res.StatusCode == http.StatusUnauthorized {
 		if err := r.a.maybeRefreshCreds(creds); err != nil {
-			klog.Errorf("refreshing credentials: %v", err)
+			klog.ErrorS(err, "refreshing credentials")
 		}
 	}
 	return res, nil
 }
 
-func (a *Authenticator) credsExpired() bool {
+// contextKey type is unexported to prevent collisions.
+type contextKey int
+
+const credsKey contextKey = iota
+
+func (a *Authenticator) credsExpiredLocked() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
-func (a *Authenticator) cert() (*tls.Certificate, error) {
-	creds, err := a.getCreds()
+func (a *Authenticator) cert(ctx context.Context) (*tls.Certificate, error) {
+	// ensure a single exec plugin call per RoundTrip
+	if credsVal := ctx.Value(credsKey); credsVal != nil {
+		creds, ok := credsVal.(*credentials)
+		if !ok {
+			return nil, fmt.Errorf("unexpected type stored in creds: %T", creds)
+		}
+		return creds.cert, nil
+	}
+
+	// fallback to fetching creds directly if we are not called via RoundTrip
+	creds, connClosed, err := a.getCreds()
 	if err != nil {
 		return nil, err
 	}
+	if connClosed {
+		// If we close the underlying TCP connection in the process of starting the TLS handshake,
+		// there is no point in letting the handshake continue.  At least this way we provide a
+		// better error message then "write tcp host:port->host:port use of closed network connection"
+		return nil, errors.New("failing TLS handshake due to certificate rotation")
```

Comment:
@liggitt see if you think 13a4745349d002f0b725545f3c9bafed443babd2 is a hack we would want to live with.

Reply:
(We will likely need something more sophisticated because if multiple TLS connections are established near the same time it could cause unrelated connections to be closed).